### PR TITLE
Fix events RestApi paging to latest

### PR DESCRIPTION
**Pull request by ikabiljo**
### Test Plan
Added unit test for it

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2582)

### CODE REVIEW

**Comment by banool** [comment]
Apologies I should probably write up a big READ_ME_PLEASE file in the API directory, but could you make this change to the v1 API as well? Or you could do just the v1 API, not the v0 API. 

In short, instead of changing `events.rs`, `page.rs`, and `transactions.rs`, change those files in the `poem_backend` directory instead.

I'll go write up that big warning note now.

File: api/src/poem_backend/events.rs [hunk 1]
```diff
@@ -87,14 +87,17 @@ impl EventsApi {
         event_key: EventKey,
     ) -> BasicResultWith404<Vec<Event>> {
         let latest_ledger_info = self.context.get_latest_ledger_info_poem()?;
+        let ledger_version = latest_ledger_info.version();
+        let last_sequence_number = self
```
Comment (msmouse) [inline api/src/poem_backend/events.rs:91]: why don't we utilize the DB interface to get the latest events?

https://github.com/aptos-labs/aptos-core/blob/c817048580bcc422a268591bd214ee66ab9dcbb1/storage/aptosdb/src/lib.rs#L1074
https://github.com/aptos-labs/aptos-core/blob/c817048580bcc422a268591bd214ee66ab9dcbb1/storage/aptosdb/src/lib.rs#L647

Comment (ikabiljo) [reply api/src/poem_backend/events.rs:91]: @msmouse :

current get_events API doesn't accept ledger_version, but we need API to be consistent about what it returns. That means that if new events are inserted from fetching latest_ledger_info to the time calling get_events, paging would be misaligned, and I would need to have more complicated logic, and issue multiple get_events calls.

So I have two options - adding ledger_version to the get_events call, or adding this new function to return last sequence number.
API is similar between get_transactions and get_events, but storage functions are not aligned. So I wanted to pick a solution that makes the API logic as similar between them as possible. Given transactions "sequence numbers" are versions, and storage provides that API to fetch it, adding API to fetch latest sequence number for event made the usage look very similar, and allowing me to reuse code (i.e. page.compute_start method I added).

Let me know if you have a different suggestion, that solves the first problem, and has consistency between the APIs?

Comment (msmouse) [reply api/src/poem_backend/events.rs:91]: https://github.com/aptos-labs/aptos-core/pull/2652 let's see if this passes the tests.

File: api/src/poem_backend/transactions.rs [hunk 2]
```diff
@@ -263,15 +263,10 @@ impl TransactionsApi {
     fn list(&self, accept_type: &AcceptType, page: Page) -> BasicResultWith404<Vec<Transaction>> {
         let latest_ledger_info = self.context.get_latest_ledger_info_poem()?;
         let ledger_version = latest_ledger_info.version();
+
         let limit = page.limit()?;
         // TODO: https://github.com/aptos-labs/aptos-core/issues/2286
-        let last_page_start = if ledger_version > (limit as u64) {
-            ledger_version - (limit as u64)
```
Comment (ikabiljo) [inline api/src/poem_backend/transactions.rs:269]: this was actually a bug, it should've been limit -1. It currently skips the most recent transaction (version==ledger_version)
I fixed this (in compute_start) and added unit test for it.

**Review by ikabiljo — COMMENTED** [review]
@msmouse This has a bit of uglier than before (due to it not being symmetric) 
let me know what do you think

File: api/src/context.rs [hunk 3]
```diff
@@ -405,17 +405,32 @@ impl Context {
...
         ledger_version: u64,
     ) -> Result<Vec<EventWithVersion>> {
-        let events = self
-            .db
-            .get_events(event_key, start, Order::Ascending, limit as u64)?;
-        Ok(events
-            .into_iter()
-            .filter(|event| event.transaction_version <= ledger_version)
-            .collect::<Vec<_>>())
+        if let Some(start) = start {
```
Comment (msmouse) [inline api/src/context.rs:412]: https://github.com/aptos-labs/aptos-core/pull/2665 

^ this'll make it clean, right?

**Comment by ikabiljo** [comment]
Discuss with @msmouse , storage interaction used here should the right one for now, he might clean up things a bit more later.

this is ready for review

File: api/src/page.rs [hunk 4]
```diff
@@ -19,6 +19,11 @@ pub(crate) struct Page {
 }
 
 impl Page {
+    pub fn compute_start(&self, limit: u16, max: u64) -> Result<u64, Error> {
```
Comment (msmouse) [inline api/src/page.rs:22]: Is `limit` checked somewhere to be non-zero?

File: crates/aptos-rest-client/src/lib.rs [hunk 5]
```diff
@@ -347,6 +349,89 @@ impl Client {
...
+        start: Option<u64>,
+        limit: Option<u64>,
+    ) -> Result<Response<Vec<Event>>> {
+        let url = self.base_url.join(&format!(
+            "accounts/{}/events/{}/{}",
+            address.to_hex_literal(),
+            struct_tag,
+            field_name
+        ))?;
+        println!("{}", url.clone());
```
Comment (msmouse) [inline crates/aptos-rest-client/src/lib.rs:366]: remove?

File: crates/aptos-rest-client/src/lib.rs [hunk 6]
```diff
@@ -347,6 +349,89 @@ impl Client {
...
+                start,
+                limit,
+            )
+            .await?;
+
+        response.and_then(|events| {
+            let new_events: Result<Vec<_>> = events
+                .into_iter()
+                .map(|event| {
+                    println!("{:?}", &event.data);
```
Comment (msmouse) [inline crates/aptos-rest-client/src/lib.rs:414]: remove?

**Review by msmouse — APPROVED** [review]

File: testsuite/smoke-test/src/client.rs [hunk 7]
```diff
@@ -106,3 +106,40 @@ async fn test_concurrent_transfers_single_node() {
...
+        .await
+        .unwrap()
+        .into_inner();
+    let start_transations = client
+        .get_transactions(None, Some(2))
+        .await
+        .unwrap()
+        .into_inner();
+
+    let _ = create_and_fund_account(&mut swarm, 100).await;
```
Comment (msmouse) [inline testsuite/smoke-test/src/client.rs:125]: `let _ = ` not necessary?

**Pull request merged**